linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala

```diff
@@ -0,0 +1,611 @@
+    b.appendElement(tagNames(tagID))
+
+  private def appendName(globalID: GlobalID)(implicit b: WatBuilder): Unit =
+    b.appendElement(globalNames(globalID))
+
+  private def appendName(typeID: TypeID, fieldID: FieldID)(implicit b: WatBuilder): Unit =
+    b.appendElement(fieldNames(typeID)(fieldID))
+
+  private def appendName(localID: LocalID)(implicit b: WatBuilder): Unit =
+    b.appendElement(localNames.get(localID))
+
+  private def appendName(labelID: LabelID)(implicit b: WatBuilder): Unit =
+    b.appendElement(labelNames.get(labelID))
+
+  private def writeModule(module: Module)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "module", {
+        module.types.foreach(writeRecType)
+        module.imports.foreach(writeImport)
+        module.funcs.foreach(writeFunction)
+        module.tags.foreach(writeTag)
+        module.globals.foreach(writeGlobal)
+        module.exports.foreach(writeExport)
+        module.start.foreach(writeStart)
+        module.elems.foreach(writeElement)
+        module.datas.foreach(writeData)
+      }
+    )
+  }
+
+  private def writeRecType(recType: RecType)(implicit b: WatBuilder): Unit = {
+    recType.subTypes match {
+      case singleSubType :: Nil =>
+        writeTypeDefinition(singleSubType)
+      case subTypes =>
+        b.newLineList(
+          "rec", {
+            subTypes.foreach(writeTypeDefinition)
+          }
+        )
+    }
+  }
+
+  private def writeTypeDefinition(subType: SubType)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "type", {
+        appendName(subType.id)
+        subType match {
+          case SubType(_, _, true, None, compositeType) =>
+            writeCompositeType(subType.id, compositeType)
+          case _ =>
+            b.sameLineList(
+              "sub", {
+                if (subType.isFinal)
+                  b.appendElement("final")
+                for (superType <- subType.superType)
+                  appendName(superType)
+                writeCompositeType(subType.id, subType.compositeType)
+              }
+            )
+        }
+      }
+    )
+  }
+
+  private def writeCompositeType(typeID: TypeID, t: CompositeType)(implicit b: WatBuilder): Unit = {
+    def writeFieldType(fieldType: FieldType): Unit = {
+      if (fieldType.isMutable)
+        b.sameLineList(
+          "mut", {
+            writeType(fieldType.tpe)
+          }
+        )
+      else writeType(fieldType.tpe)
+    }
+
+    def writeField(field: StructField): Unit = {
+      b.sameLineList(
+        "field", {
+          appendName(typeID, field.id)
+          writeFieldType(field.fieldType)
+        }
+      )
+    }
+
+    t match {
+      case FunctionType(params, results) =>
+        b.sameLineList(
+          "func", {
+            params.foreach { ty =>
+              b.sameLineList("param", writeType(ty))
+            }
+            results.foreach { ty =>
+              b.sameLineList("result", writeType(ty))
+            }
```

Comment:
```suggestion
            writeSig(params, results)
```

Alternative, remove `writeSig` (which seems unused).